File: python/packages/var/src/var/param_diff.py
"""param_diff.py — port of typescript/packages/var-core/src/param-diff.ts.

Compare a sensor's returned inline actuals against captured document values.
"""
from __future__ import annotations

from typing import Any, Sequence

from var.cell_diff import CellDiff
from var.span import Span


def compare_params(
    returned: Sequence[Any],
    expected: Sequence[Any],
    param_spans: Sequence[Span],
    source_texts: Sequence[str],
) -> tuple[CellDiff, ...]:
    """Compare returned actuals against expected values captured from the document.

    ``expected``, ``param_spans``, and ``source_texts`` align 1:1 with ``returned``;
    the caller validates length first.  Structural equality (Python ``==``) is used
    so objects compare by value across references.
    """
    diffs: list[CellDiff] = []
    for i in range(len(expected)):
        ok = returned[i] == expected[i]
        diffs.append(
            CellDiff(
                column=f"arg {i + 1}",
                span=param_spans[i],
                expected=source_texts[i] if i < len(source_texts) else str(expected[i]),
                actual=str(returned[i]),
                ok=ok,
            )
        )
    return tuple(diffs)

File: dotnet/Varar.Core/ParamDiff.cs
using System.Collections.Immutable;

namespace Varar.Core;

public static class ParamDiff
{
    /// <summary>
    /// Compare a sensor's returned inline actuals against the captured arguments. Port of
    /// <c>param-diff.ts</c>: <paramref name="sourceTexts"/> supplies the diff's expected display and
    /// <paramref name="formats"/> renders the actual side (document notation) when present.
    /// </summary>
    public static ImmutableArray<CellDiff> CompareParams(
        IReadOnlyList<Value> returned,
        ImmutableArray<Value> expected,
        ImmutableArray<Span> paramSpans,
        IReadOnlyList<string> sourceTexts,
        ImmutableArray<ParameterFormat?> formats)
    {
        var diffs = ImmutableArray.CreateBuilder<CellDiff>();
        for (int i = 0; i < expected.Length; i++)
        {
            bool ok = returned[i].Equals(expected[i]);
            var format = i < formats.Length ? formats[i] : null;
            var (actualText, viaFormat) = RenderParamValue(returned[i], format);
            var expectedText = i < sourceTexts.Count
                ? sourceTexts[i]
                : RenderParamValue(expected[i], format).Text;
            diffs.Add(new CellDiff(
                $"cell {i + 1}",
                paramSpans[i],
                expectedText,
                actualText,
                ok,
                expected[i],
                returned[i],
                viaFormat));
        }

        return diffs.ToImmutable();
    }

    private static (string Text, bool ViaFormat) RenderParamValue(Value value, ParameterFormat? format)
    {
        if (format is not null)
        {
            try
            {
                return (format(value), true);
            }
            catch
            {
                // fall through to the generic rendering
            }
        }

        return (CellDiffs.RenderCellValue(value), false);
    }
}
